packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -322,22 +397,120 @@ class TimelineEvent {
   TimelineEvent parent;
   List<TimelineEvent> children = <TimelineEvent>[];
 
-  int get duration => endTime - startTime;
-
-  bool get wellFormed => startTime != null && duration != null;
+  int get duration => (endTime != null) ? endTime - startTime : null;
 
   bool get isCpuEvent => type == TimelineEventType.cpu;
   bool get isGpuEvent => type == TimelineEventType.gpu;
 
+  bool get isPipelineProduceFlow => _hasChild('Engine::BeginFrame');
+  bool get isPipelineConsumeFlow => _hasChild('PipelineConsume');
+
+  /// Whether there is a child with the given name [childName] is contained
+  /// somewhere in the subtree [children].
+  bool _hasChild(String childName) {
+    bool childFound = false;
+
+    void findChild(TimelineEvent event, String childName) {
+      if (event.name.contains(childName)) {
+        childFound = true;
+      }
+      if (childFound) return;
+      for (TimelineEvent e in event.children) {
+        findChild(e, childName);
+      }
+    }
+
+    findChild(this, childName);
+    return childFound;
+  }
+
+  /// Adds a child event.
+  ///
+  /// If we know the child's location (knownLocation == true)
+  /// add the child directly via [_addChild]. If we do not know the child's
+  /// location, find it's proper place in the subtree and add the child there.
+  void addChild(TimelineEvent child, {bool findChildLocation = false}) {
+    bool newChildAdded = false;
+
+    // Places the child in it's correct position amongst the other children.
+    void _putChildInSubtree(TimelineEvent root, TimelineEvent newChild) {
+      // [root] is a leaf. Add child here.
+      if (root.children.isEmpty) {
+        root._addChild(child);
+        newChildAdded = true;
+      }
+
+      if (newChildAdded) return;
+
+      final List<TimelineEvent> _children =
+          List<TimelineEvent>.from(root.children);
+      for (int i = 0; i < _children.length; i++) {
+        final TimelineEvent currentChild = _children[i];
+
+        // [newChild] is the parent of [currentChild].
+        if (newChild._isParentOf(currentChild)) {
+          // Link [currentChild] with its correct parent [newChild].
+          newChild._addChild(currentChild);
+
+          // Unlink [child] from its incorrect parent [this].
```

Comment:
it seems a bit odd that we are creating a bunch of nodes with incorrect parents. it would be nice to just not assign parents and then later assign parents once we have enough information.

Reply:
We will only need to reorder children in the case of DurationComplete events. DurationBegin and DurationEnd events are guaranteed to come in order of timestamp, but there is not the same guarantee for DurationComplete events (which have both a timestamp field and a duration field).

DurationBegin/End events are already added with the simple _addChild method that performs `this.children.add(child)` and `child.parent = this`, so we are not doing any extra work there.

The work we will need to do, regardless of when we do it, is verify the DurationComplete events are added in the proper place in the tree. This code does that as we receive each event.

From looking through event traces, this is the typical problem this code solves

If events come in the order E1 - E6 where E3- E6 are duration complete events, adding them in chronological order of reception would yield this tree.
                           E1
                            |
                           E2
                               \    \    \     \
                             E3   E4  E5  E6

**Issue** - the timestamps indicate that E6 should be the sole child of E2 and the parent of E3- E5 .

When we receive E6, we will place it in the tree such that this condition holds.

                           E1
                            |
                           E2
                            |
                           E6
                               \    \    \    
                             E3   E4  E5  
                         
I'm not sure there is a benefit to re-ordering at the end instead of re-ordering as we go.

Reply:
ok sounds fine. Maybe add a comment indicating what you described here.

Reply:
done.